### PR TITLE
Add Julia v1.2 to Travis

### PR DESCRIPTION
**Pull request by blegat**

### CODE REVIEW

**Comment by blegat** [comment]
The failure appear to be due to https://github.com/JuliaLang/julia/issues/32167

**Comment by odow** [comment]
Hmm. @blegat still no luck. Are there other occurrences where this would be needed?

File: src/Bridges/Variable/Variable.jl [hunk 1]
```diff
@@ -14,6 +14,15 @@ include("map.jl")
 # Bridge optimizer bridging a specific variable bridge
 include("single_bridge_optimizer.jl")
 
+# TODO(odow): the compiler in Julia <= 1.2 (and in later versions unless
+# fixed) gets stuck compiling add_constrained_variable for some inputs. This
+# method seemed necessary to fix it.
+# See https://github.com/JuliaLang/julia/issues/32167 for more.
+@nospecialize f s
```
Comment (blegat) [inline src/Bridges/Variable/Variable.jl]: Is this line needed ?

File: src/Bridges/Variable/Variable.jl [hunk 2]
```diff
@@ -14,6 +14,15 @@ include("map.jl")
...
+function MOI.Bridges.Variable.bridge_constrained_variable(BridgeType, b, s)
+    throw(MOI.UnsupportedConstraint{MOI.SingleVariable, typeof(s)}())
```
Comment (blegat) [inline src/Bridges/Variable/Variable.jl]: `MOI.SingleVariable` should be `MOIU.variable_function_type(typeof(s))`

File: src/Bridges/Variable/Variable.jl (hunk 1 above)
Comment (odow) [reply src/Bridges/Variable/Variable.jl]: Urgh. I messed up the merge.

**Comment by odow** [comment]
Progress. We're now failing the tests rather than timing out.

**Comment by odow** [comment]
An alternative fix is to replace
https://github.com/JuliaOpt/MathOptInterface.jl/blob/2558aad1321026828e973ab9d7f097100b48f89a/src/Bridges/bridge_optimizer.jl#L866-L882
with
```Julia
@noinline function _no_inlined_bridge_constrained_variable(b, set)
    BridgeType = Variable.concrete_bridge_type(b, typeof(set))
    bridge = Variable.bridge_constrained_variable(BridgeType, b, set)
    return Variable.add_key_for_bridge(Variable.bridges(b), bridge, set)
end

function MOI.add_constrained_variable(
    b::AbstractBridgeOptimizer, set::MOI.AbstractScalarSet
)
    if is_bridged(b, typeof(set)) ||
        is_bridged(b, MOI.SingleVariable, typeof(set))
        if B.supports_bridging_constrained_variable(b, typeof(set))
            return _no_inlined_bridge_constrained_variable(b, set)
        else
            variable = MOI.add_variable(b)
            constraint = MOI.add_constraint(b, MOI.SingleVariable(variable), set)
            return variable, constraint
        end
    else
        return MOI.add_constrained_variable(b.model, set)
    end
end
```

The problem is that the compiler can't infer `BridgeType` or `bridge` (see below), so without a function barrier (that is prevented from inlining!), the code generated by `add_constrained_variable` is terrible.

The current fix works around this by defining a fallback to `Variable.bridge_constrained_variable`, but I'm not as sure why it works.
```
julia> @code_warntype _no_inlined_bridge_constrained_variable(b, MOI.GreaterThan(0.0))
Variables
  #self#::Core.Compiler.Const(_function_barrier, false)
  b::MathOptInterface.Bridges.LazyBridgeOptimizer{Model{Float64}}
  set::MathOptInterface.GreaterThan{Float64}
  BridgeType::DataType
  bridge::MathOptInterface.Bridges.Variable.VectorizeBridge{Float64,_A} where _A

Body::Tuple{MathOptInterface.VariableIndex,MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.GreaterThan{Float64}}}
1 ─       $(Expr(:meta, :noinline))
│   %2  = MathOptInterface.Bridges.Variable::Core.Compiler.Const(MathOptInterface.Bridges.Variable, false)
│   %3  = Base.getproperty(%2, :concrete_bridge_type)::Core.Compiler.Const(MathOptInterface.Bridges.Variable.concrete_bridge_type, false)
│   %4  = Main.typeof(set)::Core.Compiler.Const(MathOptInterface.GreaterThan{Float64}, false)
│         (BridgeType = (%3)(b, %4))
│   %6  = MathOptInterface.Bridges.Variable::Core.Compiler.Const(MathOptInterface.Bridges.Variable, false)
│   %7  = Base.getproperty(%6, :bridge_constrained_variable)::Core.Compiler.Const(MathOptInterface.Bridges.Variable.bridge_constrained_variable, false)
│   %8  = BridgeType::DataType
│         (bridge = (%7)(%8, b, set))
│   %10 = MathOptInterface.Bridges.Variable::Core.Compiler.Const(MathOptInterface.Bridges.Variable, false)
│   %11 = Base.getproperty(%10, :add_key_for_bridge)::Core.Compiler.Const(MathOptInterface.Bridges.Variable.add_key_for_bridge, false)
│   %12 = MathOptInterface.Bridges.Variable::Core.Compiler.Const(MathOptInterface.Bridges.Variable, false)
│   %13 = Base.getproperty(%12, :bridges)::Core.Compiler.Const(MathOptInterface.Bridges.Variable.bridges, false)
│   %14 = (%13)(b)::MathOptInterface.Bridges.Variable.Map
│   %15 = bridge::MathOptInterface.Bridges.Variable.VectorizeBridge{Float64,_A} where _A
│   %16 = (%11)(%14, %15, set)::Tuple{MathOptInterface.VariableIndex,MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.GreaterThan{Float64}}}
└──       return %16
```

**Comment by odow** [comment]
Here's a puzzle on the latest MOI master, not this branch, but on Julia 1.2.
```Julia
using MathOptInterface
const MOI = MathOptInterface

@generated function MOI.Bridges.Constraint.bridge_constraint(BridgeType, b, f, s)
    Core.println("Constraint: ") #, BridgeType, b, f, s)
    return :(error("Generated Constraint"))
end
@generated function MOI.Bridges.Variable.bridge_constrained_variable(BridgeType, b, s)
    Core.println("Variable: ") #, BridgeType, b, s)
    return :(error("Generated Variable"))
end

MOI.Utilities.@model(Model, (), (), (MOI.Nonnegatives,), (), (), (), (MOI.VectorOfVariables,), ())
MOI.supports_constraint(::Model{T}, ::Type{MOI.SingleVariable}, ::Type{MOI.GreaterThan{T}}) where {T} = false
const b = MOI.Bridges.full_bridge_optimizer(Model{Float64}(), Float64);

@time MOI.add_constrained_variable(b, MOI.GreaterThan(0.0))
```
yields
```Julia
Variable: 
Variable: 
Variable: 
Variable: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Variable: 
Variable: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Constraint: 
Variable: 
Constraint: 
Variable: 
Constraint: 
Constraint: 
Variable: 
  3.272182 seconds (15.30 M allocations: 822.902 MiB, 8.54% gc time)
(MathOptInterface.VariableIndex(-1), MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.GreaterThan{Float64}}(-1))
```

But, if I try
```Julia
using MathOptInterface
const MOI = MathOptInterface

@generated function MOI.Bridges.Constraint.bridge_constraint(BridgeType, b, f, s)
    Core.println("Constraint: ", BridgeType, b, f, s)
    return :(error("Generated Constraint"))
end
@generated function MOI.Bridges.Variable.bridge_constrained_variable(BridgeType, b, s)
    Core.println("Variable: ", BridgeType, b, s)
    return :(error("Generated Variable"))
end

MOI.Utilities.@model(Model, (), (), (MOI.Nonnegatives,), (), (), (), (MOI.VectorOfVariables,), ())
MOI.supports_constraint(::Model{T}, ::Type{MOI.SingleVariable}, ::Type{MOI.GreaterThan{T}}) where {T} = false
const b = MOI.Bridges.full_bridge_optimizer(Model{Float64}(), Float64);

@time MOI.add_constrained_variable(b, MOI.GreaterThan(0.0))
```
I get 
```Julia
  3.194577 seconds (15.28 M allocations: 821.955 MiB, 8.54% gc time)
(MathOptInterface.VariableIndex(-1), MathOptInterface.ConstraintIndex{MathOptInterface.SingleVariable,MathOptInterface.GreaterThan{Float64}}(-1))
```

Why are the generated methods not being called if they print the argument types?

**Pull request merged**